File: apps/ios/Voiid/Voiid/Games/Carrom/CarromEngine.swift
//
//  CarromEngine.swift
//  Voiid Ui
//
//  Carrom rules engine, board layout generator, turn/foul/cover evaluation,
//  and intelligent raycasting Bot AI.
//

import CoreGraphics
import Foundation

// MARK: - Game Mode & Rules Configuration

enum CarromGameMode: String, CaseIterable, Identifiable {
    case classic = "Classic"

    var id: String { rawValue }

    var description: String {
        switch self {
        case .classic: return "You play white. The bot plays black. Cover the queen and clear your colour."
        }
    }
}

enum CarromTurn: Int, CaseIterable, Identifiable {
    case player1 = 0
    case player2 = 1

    var id: Int { rawValue }

    var opposite: CarromTurn {
        self == .player1 ? .player2 : .player1
    }
}

// MARK: - Engine

enum CarromEngine {

    // MARK: - Initial Board Layout

    /// Generates standard 19-piece carrom rosette at board center + striker
    /// Uses authentic Hexagonal Close-Packed (HCP) geometry where every piece is tangent
    static func generateStartingPieces() -> [CarromPiece] {
        var pieces: [CarromPiece] = []
        let center = CGPoint(x: CarromTheme.surfaceSize / 2, y: CarromTheme.surfaceSize / 2) // (180, 180)
        let r = CarromTheme.pieceRadius // 11.5
        var idCounter = 1

        // 1. Center Queen
        pieces.append(CarromPiece(
            id: idCounter,
            type: .queen,
            position: center,
            radius: r,
            mass: CarromTheme.pieceMass
        ))
        idCounter += 1

        // 2. Inner Ring (6 pieces: 3 White, 3 Black alternating)
        // Tangent to queen at distance 2 * r
        let innerDistance = r * 2.0
        for i in 0..<6 {
            let angle = CGFloat(i) * (.pi / 3.0) // 60 deg intervals
            let pos = CGPoint(
                x: center.x + cos(angle) * innerDistance,
                y: center.y + sin(angle) * innerDistance
            )
            let type: CarromPieceType = (i % 2 == 0) ? .white : .black
            pieces.append(CarromPiece(
                id: idCounter,
                type: type,
                position: pos,
                radius: r,
                mass: CarromTheme.pieceMass
            ))
            idCounter += 1
        }

        // 3. Outer Ring (12 pieces in HCP arrangement: 6 corners at 4r, 6 edges at 2*sqrt(3)*r)
        // Alternates White and Black to give 6 White and 6 Black in outer ring (total 9W, 9B, 1Q)
        let cornerDistance = r * 4.0
        let edgeDistance = 2.0 * sqrt(3.0) * r

        var outerSlots: [(pos: CGPoint, type: CarromPieceType)] = []
        for i in 0..<6 {
            let cornerAngle = CGFloat(i) * (.pi / 3.0)
            let cornerPos = CGPoint(
                x: center.x + cos(cornerAngle) * cornerDistance,
                y: center.y + sin(cornerAngle) * cornerDistance
            )
            // Pattern: corner is opposite of inner piece at same angle
            let cornerType: CarromPieceType = (i % 2 == 0) ? .black : .white
            outerSlots.append((pos: cornerPos, type: cornerType))

            let edgeAngle = cornerAngle + (.pi / 6.0)
            let edgePos = CGPoint(
                x: center.x + cos(edgeAngle) * edgeDistance,
                y: center.y + sin(edgeAngle) * edgeDistance
            )
            let edgeType: CarromPieceType = (i % 2 == 0) ? .white : .black
            outerSlots.append((pos: edgePos, type: edgeType))
        }

        for slot in outerSlots {
            pieces.append(CarromPiece(
                id: idCounter,
                type: slot.type,
                position: slot.pos,
                radius: r,
                mass: CarromTheme.pieceMass
            ))
            idCounter += 1
        }

        // 4. Initial Striker on Player 1's baseline
        let strikerPos = baselinePosition(for: .player1, normalizedX: 0.5)
        pieces.append(CarromPiece(
            id: 0,
            type: .striker,
            position: strikerPos,
            radius: CarromTheme.strikerRadius,
            mass: CarromTheme.strikerMass
        ))

        return pieces
    }

    // MARK: - Baseline Geometry

    /// Baseline range: left X = 74, right X = 286
    static let baselineMinX: CGFloat = 74.0
    static let baselineMaxX: CGFloat = CarromTheme.surfaceSize - 74.0

    static func baselineY(for turn: CarromTurn) -> CGFloat {
        if turn == .player1 {
            return CarromTheme.surfaceSize - CarromTheme.baselineInset // Bottom: 308
        } else {
            return CarromTheme.baselineInset // Top: 52
        }
    }

    static func baselinePosition(for turn: CarromTurn, normalizedX: CGFloat) -> CGPoint {
        let clampedX = min(max(normalizedX, 0.0), 1.0)
        let x = baselineMinX + (baselineMaxX - baselineMinX) * clampedX
        let y = baselineY(for: turn)
        return CGPoint(x: x, y: y)
    }

    /// Validates striker doesn't overlap any piece on the board
    static func isStrikerPlacementValid(at pos: CGPoint, pieces: [CarromPiece]) -> Bool {
        let strikerRadius = CarromTheme.strikerRadius
        for p in pieces where p.type != .striker && !p.isPocketed {
            let dist = pos.distance(to: p.position)
            if dist < (strikerRadius + p.radius + 1.0) {
                return false // Overlapping
            }
        }
        return true
    }

    // MARK: - Bot AI

    struct BotShot {
        let baselineX: CGFloat       // Normalized 0.0 ... 1.0
        let aimDirection: CGVector   // Vector to strike
        let power: CGFloat           // 300 ... 800 pt/s
    }

    /// Evaluates board and generates intelligent bot strike
    static func calculateBotShot(pieces: [CarromPiece], difficulty: Int = 2) -> BotShot {
        let pockets = [
            CGPoint(x: CarromTheme.pocketInset, y: CarromTheme.pocketInset),
            CGPoint(x: CarromTheme.surfaceSize - CarromTheme.pocketInset, y: CarromTheme.pocketInset),
            CGPoint(x: CarromTheme.surfaceSize - CarromTheme.pocketInset, y: CarromTheme.surfaceSize - CarromTheme.pocketInset),
            CGPoint(x: CarromTheme.pocketInset, y: CarromTheme.surfaceSize - CarromTheme.pocketInset)
        ]

        let availablePieces = pieces.filter { ($0.type == .black || $0.type == .queen) && !$0.isPocketed }
        guard !availablePieces.isEmpty else {
            return BotShot(baselineX: 0.5, aimDirection: CGVector(dx: 0, dy: 1), power: 500)
        }

        var bestTarget: CarromPiece? = nil
        var bestPocket: CGPoint = pockets[2]
        var bestBaselineNormX: CGFloat = 0.5
        var bestScore: CGFloat = -1000

        // Prioritize Queen if on board, then White / Black
        let testPositions = stride(from: 0.1, through: 0.9, by: 0.1).map { CGFloat($0) }

        for target in availablePieces {
            let priorityWeight: CGFloat = (target.type == .queen) ? 1.8 : 1.0

            for pocket in pockets {
                let toPocket = (pocket - target.position).normalized()
                // Ghost striker position behind the target piece
                let ghostPos = target.position - (toPocket * (CarromTheme.strikerRadius + target.radius))

                for normX in testPositions {
                    let strikerPos = baselinePosition(for: .player2, normalizedX: normX)

                    // Striker must shoot DOWNWARDS toward bottom of board
                    let shotVec = ghostPos - strikerPos
                    if shotVec.y < 20 { continue } // Can't shoot backwards

                    let shotDist = shotVec.length
                    let pocketDist = target.position.distance(to: pocket)
                    let alignment = shotVec.normalized().dot(with: toPocket)

                    // Scoring heuristic
                    let score = (alignment * 200.0) - (shotDist * 0.4) - (pocketDist * 0.3) + (priorityWeight * 60.0)

                    if score > bestScore && isStrikerPlacementValid(at: strikerPos, pieces: pieces) {
                        bestScore = score
                        bestTarget = target
                        bestPocket = pocket
                        bestBaselineNormX = normX
                    }
                }
            }
        }

        guard let target = bestTarget else {
            // Fallback shot toward center
            let freeX = stride(from: 0.0, through: 1.0, by: 0.01).map { CGFloat($0) }.first {
                isStrikerPlacementValid(at: baselinePosition(for: .player2, normalizedX: $0), pieces: pieces)
            } ?? 0.5
            let strikerPos = baselinePosition(for: .player2, normalizedX: freeX)
            let center = CGPoint(x: CarromTheme.surfaceSize / 2, y: CarromTheme.surfaceSize / 2)
            let dir = (center - strikerPos).normalized().vector
            return BotShot(baselineX: freeX, aimDirection: dir, power: 550)
        }

        let strikerPos = baselinePosition(for: .player2, normalizedX: bestBaselineNormX)
        let toPocket = (bestPocket - target.position).normalized()
        let ghostPos = target.position - (toPocket * (CarromTheme.strikerRadius + target.radius))
        var aimDir = (ghostPos - strikerPos).normalized().vector

        // Apply realistic humanized variance according to difficulty
        let jitterScale: CGFloat = (difficulty == 1) ? 0.07 : (difficulty == 2) ? 0.03 : 0.008
        let jitterAngle = CGFloat.random(in: -jitterScale...jitterScale)
        let cosA = cos(jitterAngle), sinA = sin(jitterAngle)
        aimDir = CGVector(dx: aimDir.dx * cosA - aimDir.dy * sinA,
                          dy: aimDir.dx * sinA + aimDir.dy * cosA)

        let power = CGFloat.random(in: 560...720)
        return BotShot(baselineX: bestBaselineNormX, aimDirection: aimDir, power: power)
    }
}

File: apps/android/app/src/main/java/com/voiid/app/main/games/carrom/CarromEngine.kt
package com.voiid.app.main.games.carrom

import androidx.compose.ui.graphics.Color
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sin
import kotlin.math.sqrt

/** Port of iOS `CarromTheme.swift`: geometry on the virtual 360 surface, and the palette. */
object CarromTheme {
    const val SURFACE = 360f
    const val FRAME = 24f
    const val TOTAL = SURFACE + FRAME * 2   // 408
    const val POCKET_RADIUS = 17.5f
    const val POCKET_INSET = 20f
    const val PIECE_RADIUS = 11.5f
    const val STRIKER_RADIUS = 16.5f
    const val PIECE_MASS = 1f
    const val STRIKER_MASS = 3.2f
    const val BASELINE_INSET = 52f
    const val BASELINE_WIDTH = 26f
    const val BASELINE_CIRCLE_RADIUS = 10.5f
    const val CENTER_OUTER = 36f
    const val CENTER_INNER = 14f

    val woodLight = Color(0xFFF4E7D3); val woodMid = Color(0xFFEAD8BC); val woodDark = Color(0xFFDFCAAB)
    val boardLine = Color(0xFF563721).copy(alpha = 0.85f)
    val boardLineLo = Color(0xFF563721).copy(alpha = 0.35f)
    val baseCircleRed = Color(0xFFC62828)
    val frameWoodLight = Color(0xFF361E13); val frameWoodDark = Color(0xFF140A06)
    val cushionRubber = Color(0xFF2E180E)
    val pocketWell = Color(0xFF09090B)
    val brassHighlight = Color(0xFFF5D77F); val brassBase = Color(0xFFC9A227); val brassShadow = Color(0xFF785E0E)
    val whiteMain = Color(0xFFFBF7EF); val whiteRing = Color(0xFFE2D3BA); val whiteCore = Color(0xFFC8B494)
    val blackMain = Color(0xFF221F1C); val blackRing = Color(0xFF3B3733); val blackCore = Color(0xFF151312)
    val queenMain = Color(0xFFC62828); val queenRing = Color(0xFFE53935); val queenStar = Color(0xFFFFD54F)
    val strikerBody = Color(0xFF0F172A); val strikerGlow = Color(0xFF00E5FF)
    val strikerRing = Color(0xFF38BDF8); val strikerCore = Color(0xFFE0F2FE)
    val aimLaser = Color(0xFF00F0FF)
    val aimBounceLaser = Color(0xFF38BDF8).copy(alpha = 0.65f)
    val aimGhostDisc = Color(0xFF00F0FF).copy(alpha = 0.3f)
}

enum class CarromTurn { PLAYER1, PLAYER2;
    val opposite get() = if (this == PLAYER1) PLAYER2 else PLAYER1
}

/** Port of iOS `CarromEngine`: the rosette, baseline geometry, and the raycasting bot. */
object CarromEngine {

    /** The 19-piece hexagonal close-packed rosette (queen, 9 white, 9 black) plus the striker. */
    fun generateStartingPieces(): MutableList<CarromPiece> {
        val pieces = mutableListOf<CarromPiece>()
        val c = Vec(CarromTheme.SURFACE / 2, CarromTheme.SURFACE / 2)
        val r = CarromTheme.PIECE_RADIUS
        var id = 1
        fun add(type: CarromPieceType, pos: Vec) {
            pieces += CarromPiece(id++, type, pos, radius = r, mass = CarromTheme.PIECE_MASS)
        }
        add(CarromPieceType.QUEEN, c)
        for (i in 0 until 6) {
            val a = (i * PI / 3).toFloat()
            add(if (i % 2 == 0) CarromPieceType.WHITE else CarromPieceType.BLACK, c + Vec.fromAngle(a, r * 2))
        }
        val edge = 2f * sqrt(3f) * r
        for (i in 0 until 6) {
            val a = (i * PI / 3).toFloat()
            add(if (i % 2 == 0) CarromPieceType.BLACK else CarromPieceType.WHITE, c + Vec.fromAngle(a, r * 4))
            add(if (i % 2 == 0) CarromPieceType.WHITE else CarromPieceType.BLACK, c + Vec.fromAngle(a + (PI / 6).toFloat(), edge))
        }
        pieces += CarromPiece(0, CarromPieceType.STRIKER, baselinePosition(CarromTurn.PLAYER1, 0.5f),
            radius = CarromTheme.STRIKER_RADIUS, mass = CarromTheme.STRIKER_MASS)
        return pieces
    }

    const val BASELINE_MIN_X = 74f
    const val BASELINE_MAX_X = CarromTheme.SURFACE - 74f

    fun baselineY(turn: CarromTurn) =
        if (turn == CarromTurn.PLAYER1) CarromTheme.SURFACE - CarromTheme.BASELINE_INSET else CarromTheme.BASELINE_INSET

    fun baselinePosition(turn: CarromTurn, normX: Float): Vec {
        val x = BASELINE_MIN_X + (BASELINE_MAX_X - BASELINE_MIN_X) * min(max(normX, 0f), 1f)
        return Vec(x, baselineY(turn))
    }

    fun isStrikerPlacementValid(pos: Vec, pieces: List<CarromPiece>): Boolean =
        pieces.none { it.type != CarromPieceType.STRIKER && !it.isPocketed &&
            pos.distance(it.position) < CarromTheme.STRIKER_RADIUS + it.radius + 1f }

    class BotShot(val baselineX: Float, val aimDirection: Vec, val power: Float)

    /** [difficulty] 1…3 sets the aim jitter, exactly as iOS (0.07 / 0.03 / 0.008 rad). */
    fun calculateBotShot(pieces: List<CarromPiece>, difficulty: Int = 2): BotShot {
        val pockets = listOf(
            Vec(CarromTheme.POCKET_INSET, CarromTheme.POCKET_INSET),
            Vec(CarromTheme.SURFACE - CarromTheme.POCKET_INSET, CarromTheme.POCKET_INSET),
            Vec(CarromTheme.SURFACE - CarromTheme.POCKET_INSET, CarromTheme.SURFACE - CarromTheme.POCKET_INSET),
            Vec(CarromTheme.POCKET_INSET, CarromTheme.SURFACE - CarromTheme.POCKET_INSET),
        )
        val targets = pieces.filter { (it.type == CarromPieceType.BLACK || it.type == CarromPieceType.QUEEN) && !it.isPocketed }
        if (targets.isEmpty()) return BotShot(0.5f, Vec(0f, 1f), 500f)

        var bestTarget: CarromPiece? = null
        var bestPocket = pockets[2]
        var bestX = 0.5f
        var bestScore = -1000f
        val xs = (1..9).map { it / 10f }
        for (t in targets) {
            val weight = if (t.type == CarromPieceType.QUEEN) 1.8f else 1f
            for (pocket in pockets) {
                val toPocket = (pocket - t.position).normalized()
                val ghost = t.position - toPocket * (CarromTheme.STRIKER_RADIUS + t.radius)
                for (nx in xs) {
                    val sp = baselinePosition(CarromTurn.PLAYER2, nx)
                    val shot = ghost - sp
                    if (shot.y < 20) continue   // the bot shoots DOWN the board
                    val score = shot.normalized().dot(toPocket) * 200f - shot.length * 0.4f -
                        t.position.distance(pocket) * 0.3f + weight * 60f
                    if (score > bestScore && isStrikerPlacementValid(sp, pieces)) {
                        bestScore = score; bestTarget = t; bestPocket = pocket; bestX = nx
                    }
                }
            }
        }
        val target = bestTarget ?: run {
            val freeX = (0..100).map { it / 100f }.firstOrNull {
                isStrikerPlacementValid(baselinePosition(CarromTurn.PLAYER2, it), pieces)
            } ?: 0.5f
            val sp = baselinePosition(CarromTurn.PLAYER2, freeX)
            return BotShot(freeX, (Vec(CarromTheme.SURFACE / 2, CarromTheme.SURFACE / 2) - sp).normalized(), 550f)
        }
        val sp = baselinePosition(CarromTurn.PLAYER2, bestX)
        val toPocket = (bestPocket - target.position).normalized()
        val ghost = target.position - toPocket * (CarromTheme.STRIKER_RADIUS + target.radius)
        var aim = (ghost - sp).normalized()
        val jitter = when (difficulty) { 1 -> 0.07f; 2 -> 0.03f; else -> 0.008f }
        val a = (Math.random().toFloat() * 2 - 1) * jitter
        aim = Vec(aim.x * cos(a) - aim.y * sin(a), aim.x * sin(a) + aim.y * cos(a))
        return BotShot(bestX, aim, 560f + Math.random().toFloat() * 160f)
    }

}
